extensions/big-data-cluster/src/bigDataCluster/controller/clusterControllerApi.ts

```diff
@@ -0,0 +1,105 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { EndpointRouterApi } from './apiGenerated';
+
+export async function getEndPoints(
+	url: string, username: string, password: string, ignoreSslVerification?: boolean
+): Promise<IEndPointsResponse> {
+
+	if (!url || !username || !password) {
+		return undefined;
+	}
+
+	url = adjustUrl(url);
+	let ep = new EndpointRouterApi(username, password, url);
+	ep.ignoreSslVerification = !!ignoreSslVerification;
+
+	let controllerResponse: IEndPointsResponse = undefined;
+	let controllerError: IControllerError = undefined;
+	let request = <IEndPointsRequest>{
+		url: url,
+		username: username,
+		password: password,
+		method: 'endPointsGet'
+	};
+
+	try {
+		let result = await ep.endpointsGet();
+		controllerResponse = <IEndPointsResponse>{
+			response: result.response as IHttpResponse,
+			endPoints: result.body as IEndPoint[],
+			request
+		};
+		return controllerResponse;
+	} catch (error) {
+		if ('response' in error) {
+			let err: IEndPointsResponse = error as IEndPointsResponse;
+			let errCode = `${err.response.statusCode || ''}`;
+			let errMessage = err.response.statusMessage;
+			let errUrl = err.response.url;
+			controllerError = <IControllerError>{
+				address: errUrl,
+				code: errCode,
+				errno: errCode,
+				message: errMessage,
+				name: undefined
+			};
+		} else {
+			controllerError = error as IControllerError;
+		}
+		throw Object.assign(controllerError, { request }) as IControllerError;
```

Comment:
How are we using the request? In particular, can I verify the password is not logged anywhere or printed?

Reply:
The request object is being used for adding controller in extension tree. But there is a way not to use the request object for adding. I will remove the request object from response in next PR.